### PR TITLE
implement the `?` operator

### PR DESCRIPTION
**Pull request by japaric**
The `?` postfix operator is sugar equivalent to the try! macro, but is more amenable to chaining:
`File::open("foo")?.metadata()?.is_dir()`.

`?` is accepted on any _expression_ that can return a `Result`, e.g. `x()?`, `y!()?`, `{z}?`,
`(w)?`, etc. And binds more tightly than unary operators, e.g. `!x?` is parsed as `!(x?)`.

cc #31436

---

cc @aturon @eddyb

### CODE REVIEW

File: src/librustc_front/lowering.rs [hunk 1]
```diff
@@ -1530,6 +1530,62 @@ pub fn lower_expr(lctx: &LoweringContext, e: &Expr) -> P<hir::Expr> {
...
+                            expr_call(lctx, e.span, from, hir_vec![err_expr], None)
+                        };
+                        let err_expr = {
+                            let path = std_path(lctx, &["result", "Result", "Err"]);
+                            let path = path_global(e.span, path);
+                            let err_ctor = expr_path(lctx, path, None);
+                            expr_call(lctx, e.span, err_ctor, hir_vec![from_expr], None)
+                        };
+                        let err_pat = pat_err(lctx, e.span, pat_ident(lctx, e.span, err_ident));
+                        let ret_expr = expr(lctx, e.span, hir::Expr_::ExprRet(Some(err_expr)), None);
```
Comment (oli-obk) [inline src/librustc_front/lowering.rs]: this line is too long for `make tidy`

**Comment by durka** [comment]
Needs a feature gate, right?

**Comment by japaric** [comment]
`?` is now behind a feature gate.

**Comment by brson** [comment]
:tada:

**Comment by brson** [comment]
It might be useful to have tests of what happens when the try expansion fails to resolve. Is it nice? Hideous?

**Comment by brson** [comment]
[Here's the `try` test case](https://github.com/japaric/rust/blob/rfc243/src/test/run-pass/try-operator.rs). I think it looks nice. The `?` really gets out of the way so you don't notice the error handling.

**Comment by seanmonstar** [comment]
The RFC claims this feature flag should be `quedtion_mark`. https://github.com/rust-lang/rfcs/blob/master/text/0243-trait-based-exception-handling.md

**Comment by japaric** [comment]
> It might be useful to have tests of what happens when the try expansion fails to resolve. Is it nice? Hideous?

I don't have a build at hand, but the error message should be similar to the one for unresolved `for` loops. The one for `for` loops looks like this:

``` rust
#![feature(no_core)]
#![no_core]

fn main() {
    for x in xs {}
}
```

```
$ rustc for.rs
for.rs:5:14: 5:16 help: run `rustc --explain E0425` to see a detailed explanation
for.rs:5:5: 5:9 error: failed to resolve. Maybe a missing `extern crate iter`? [E0433]
for.rs:5     for x in xs {}
             ^~~~
for.rs:5:5: 5:9 help: run `rustc --explain E0433` to see a detailed explanation
for.rs:5:5: 5:19 error: unresolved name `iter::IntoIterator::into_iter` [E0425]
for.rs:5     for x in xs {}
             ^~~~~~~~~~~~~~
for.rs:5:5: 5:19 help: run `rustc --explain E0425` to see a detailed explanation
for.rs:5:5: 5:9 error: failed to resolve. Maybe a missing `extern crate iter`? [E0433]
for.rs:5     for x in xs {}
             ^~~~
for.rs:5:5: 5:9 help: run `rustc --explain E0433` to see a detailed explanation
for.rs:5:5: 5:19 error: unresolved name `iter::Iterator::next` [E0425]
for.rs:5     for x in xs {}
             ^~~~~~~~~~~~~~
for.rs:5:5: 5:19 help: run `rustc --explain E0425` to see a detailed explanation
for.rs:5:5: 5:19 error: unresolved enum variant, struct or const `Some` [E0419]
for.rs:5     for x in xs {}
             ^~~~~~~~~~~~~~
for.rs:5:5: 5:19 help: run `rustc --explain E0419` to see a detailed explanation
for.rs:5:5: 5:19 error: unresolved enum variant, struct or const `None` [E0419]
for.rs:5     for x in xs {}
             ^~~~~~~~~~~~~~
```

@seanmonstar will push a commit renaming the feature gate in a bit.

**Comment by nikomatsakis** [comment]
r? @nikomatsakis

(stealing review)

**Comment by nikomatsakis** [comment]
So first review done, comments:

- [ ] Move parsing code into `parse_dot_or_call_expr_with_` instead
- [ ] Move feature-gate code into `PostExpansionVisitor`
- [ ] Feels like there could/should be some negative tests, but I'm not sure precisely what :) maybe not needed

**Comment by japaric** [comment]
> Move parsing code into parse_dot_or_call_expr_with_ instead
> Move feature-gate code into PostExpansionVisitor 

Done. The impl looks simpler now. :-)

> Feels like there could/should be some negative tests, but I'm not sure precisely what :) maybe not needed

You mean like checking that `ExprKind::Try(ExprKind::Binary)` is never created?

> This seems to be roughly what we do elsewhere -- but do we need to do anything to make hygiene work out for us? Can we make a test where the context defines const val: u32 = 0 or something, just to make sure that the name val here is always a binding and does not conflict with this in-scope constant?

I've added a test for this.

**Comment by japaric** [comment]
> :umbrella: The latest upstream changes (presumably #30884) made this pull request unmergeable. Please resolve the merge conflicts.

Will rebase after @nikomatsakis takes a second look :-).

**Comment by nikomatsakis** [comment]
@japaric 

The changes look great. r=me, feel free to rebase. :)

> You mean like checking that ExprKind::Try(ExprKind::Binary) is never created?

No, I meant some kind of test that `a + b?` parses as expected, but it doesn't seem necessary now with the newer code.

**Comment by nikomatsakis** [comment]
@japaric some error messages need correcting: http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/8286/steps/test/logs/stdio

**Comment by perlun** [comment]
For the record, this was stabilized as part of [1.13.0](https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1130-2016-11-10), i.e no longer behind a feature gate. I hadn't seen it until now, when I happened to see it being referred to in [the docs](https://doc.rust-lang.org/std/fs/struct.File.html). 😉 Nice with some more syntactic sugar!

**Review by castroed47 — APPROVED** [review]

**Pull request merged**